Implement ctile

jsha opened the pull request:
_nota bene: land this change using the "rebase" method, so that these initial commits are preserved_

jsha committed:
Initial commit

jsha committed:
Add README

jsha committed:
cbor and gzip

jsha committed:
Update README

jsha committed:
close gzip writer

jsha committed:
Add MPL license

jmhodges committed:
weave Context through the server (#1)

This adds context.Context to the various client calls made by the proxy
to enable deadlining, cancellation, and other handy things. This doesn't
choose a top-level request timeout, but allows for one to be set in
another PR.

jmhodges committed:
add timeouts for the server and requests (#2)

Using the previous context code, apply some timeouts to ensure the HTTP
server stays up under traffic.

This weaves a default 4 second per-request timeout through the handler.
This timeout covers how long it takes the HTTP handler to return data
from S3 (if it exists) or to discover the tile isn't in S3, fetch the
tile data from S3 the backend, and write it before returning. This is
configurable.

This also sets the basic HTTP server timeouts that are required for
healthy Go HTTP servers. Those aren't configurable.

A good summary of Go's HTTP server timeouts and why they're needed can
be found at https://blog.cloudflare.com/exposing-go-on-the-internet/

The `TimeoutHandler` sets the timeout on the `http.Request`'s context,
so we can use `r.Context()` in the handler to propagate the deadline to
the S3 and HTTP clients.

The values for the timeouts were chosen arbitrarily but from values that
I understand to be safe for public consumption. The
`full-request-timeout` default is the one most likely to need a change
depending on how far from S3 this server is hosted (or if the work to
write to S3 is moved to a background goroutine or whatever).

Future work: Given that we're using `http.Server` directly now, it might
be time to break out the handler func as a top-level function but I
leave that for a possible follow-up.

jsha committed:
allow multiple backends per S3 bucket

Qualify the S3 key by the backend URL and the tile size.

Also, keep track of the backend URL in the `tile` object.

jsha committed:
Pass through error codes from backend

Also add -s3-prefix override (defaults to -backend)

jsha committed:
document partial tile behavior

jsha committed:
Remove cbor as output format

jsha committed:
Rename backend to logURL

jsha committed:
Document example invocation

jmhodges committed:
add tileCachingHandler to break out handler func (#3)

This moves the HTTP code out of the main func and into its own type
called `tileCachingHandler`. It's a simple refactoring that adds some
fields to the new struct to keep the pattern of adding flags in main.

jsha committed:
Update to v2 of aws-sdk-go

jsha committed:
aws-sdk v2: fix error type handling

jsha committed:
test: add integration test

aarongable commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,21 @@
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
```
I like having this run on `branches: '**'` so that PRs targeting other PR branches (i.e. stacked changes) still get CI run.

aarongable commented on main.go:
```diff
@@ -0,0 +1,367 @@
+package main
```
hyper nit: package comment?

aarongable commented on main.go:
```diff
@@ -0,0 +1,367 @@
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+
```
hyper-nit: fxamacker isn't a package local to this module, so it should be imported in the same block as the other third-party imports
```suggestion
```

aarongable commented on main.go:
```diff
@@ -0,0 +1,367 @@
+	}
+	startInt, err := strconv.ParseInt(start, 10, 64)
+	if err != nil || startInt < 0 {
+		return 0, 0, errors.New("invalid start parameter")
```
optionally:
```suggestion
		return 0, 0, fmt.Errorf("invalid start parameter: %w", err)
```

aarongable commented on main.go:
```diff
@@ -0,0 +1,367 @@
+}
+
+// tile represents important numbers about a tile: where it starts, where it ends, its size,
+// and what CT backend URL it exists on (or is anticipated to exist on), and what s3 prefix
```
```suggestion
// what CT backend URL it exists on (or is anticipated to exist on), and what s3 prefix
```

aarongable commented on main.go:
```diff
@@ -0,0 +1,367 @@
+// it should be stored/retrieved under.
+type tile struct {
+	start    int64
+	end      int64
```
Probably worth being explicit in the struct comment that `end` is an _exclusive_ bound.

jsha replied:
oop, I documented that this is a half-open interval "like CT" but then I realized CT uses a closed interval (will return the entry for `end`). Gonna check some code real quick..

jsha replied:
Alright, updated the comments and implementation. The current vibe is that half-open intervals are better (easier to reason about, fewer +1/-1 adjustments), so we use them internally, and only convert to and from closed intervals when parsing the query and when generating the outbound URL for requests. Open to feedback that it would be simpler the other way around (i.e. just use closed intervals internally because that's what CT does).

aarongable commented on main.go:
```diff
@@ -0,0 +1,367 @@
+		}
+
+		w.Header().Set("X-Source", "CT log")
+	} else if err != nil {
```
nit: I'd move this case above the getTileFromBackend case, so that the simple "just return the error" case isn't squashed between the "kinda-success" and "complete-success" cases.

aarongable commented on main.go:
```diff
@@ -0,0 +1,367 @@
+// If the backend returns a non-200 status code, it returns a statusCodeError,
+// so the caller can handle that case specially by propagating the backend's
+// status code (for instance, 400 or 404).
+func getTileFromBackend(ctx context.Context, path string, t tile) (*entries, error) {
```
It's interesting to me that all three of these big helpers (getTileFromBackend, getFromS3, and writeToS3) _could_ be methods on the tile struct. Is there a reason you chose not to go that route? (Not saying you should, just curious.)

jsha replied:
Path dependence, really. The `tile` struct came later; at first each of the big helpers had a pile o' parameters. As I abstracted them into a common struct it never occurred to me to switch them to being methods.

aarongable commented on main.go:
```diff
@@ -0,0 +1,367 @@
+	start    int64
+	end      int64
+	size     int64
+	logURL   string
+	s3prefix string
```
It feels a little weird to me that a tile carries these two pieces of data (which are identical for all tiles within a single log), but doesn't carry the s3 bucket or the API path under that log hostname, which are also identical for all tiles within a single log.

It results in odd function signatures below, where both getFromS3 and writeToS3 take a bucket name but then rely on the tile for the prefix, despite both of those ultimately coming from the same CLI flags.

I don't think I'm advocating for removing these struct memebers -- replacing them with function args everywhere would be annoyingly verbose. So maybe the s3bucket should also be a struct member here? Or maybe things are fine as they are. Just worth thinking about.

_edit_: Or maybe, in combination with my comment above about the various helpers being methods, they could be methods _on the handler struct_, so that the logURL, bucket, and prefix are all accessed from those pre-existing struct members?

jsha replied:
yeah I wound up moving the helpers onto the handler struct. good idea. 👍🏻

that allowed me to remove the `s3prefix` from `tile`, which was nice. I felt a little awkward having `tile` represent both data about the CT log and about the S3 storage, like it was trying to be in two worlds at once. Now it just contains data about the CT log, which is nice.

aarongable commented on main.go:
```diff
@@ -0,0 +1,367 @@
+// so the caller can handle that case specially by propagating the backend's
+// status code (for instance, 400 or 404).
+func getTileFromBackend(ctx context.Context, path string, t tile) (*entries, error) {
+	url := fmt.Sprintf("%s/%s?start=%d&end=%d", t.logURL, path, t.start, t.end)
```
This could be a method `t.url(path)` just like `t.key()`. I realize it's only used in the one place, while key() is used in both the get and write methods, so maybe it's not worth it, but the symmetry might be nice.

aarongable commented on main.go:
```diff
@@ -0,0 +1,367 @@
+	url := fmt.Sprintf("%s/%s?start=%d&end=%d", t.logURL, path, t.start, t.end)
+	r, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create backend Request object: %s", err)
```
```suggestion
		return nil, fmt.Errorf("unable to create backend Request object: %w", err)
```
more instances of this below, too

aarongable reviewed (commented):
Generally LGTM, I didn't spot any bugs. Just a few nits and organizational comments/questions.

Preston12321 commented on README.md:
```diff
@@ -0,0 +1,59 @@
+protocol allows the server to return fewer results than requested, so CTile does
+not attempt to request multiple tiles to fulfil a large request. If a request's
+`start` parameter is one less than the end of a tile, CTile will respond with a
+single entry. This is similar to how Trillian's [align_getentries
+flag](https://github.com/google/certificate-transparency-go/blob/6e118585d9d9757b739353829becec378f47e10b/trillian/ctfe/handlers.go#L50)
+works, and is in fact compatible with that flag, so long as CTile's tile size is
+equal to Trillian's max_get_entries flag.
```
```suggestion
single entry. This is similar to how Trillian's [align_getentries
flag](https://github.com/google/certificate-transparency-go/blob/6e118585d9d9757b739353829becec378f47e10b/trillian/ctfe/handlers.go#L50)
works, and is in fact compatible with that flag, so long as CTile's tile size is
less than or equal to Trillian's max_get_entries flag.
```

The one simplification that CTile makes to the logic is that you can only receive entries from the tile containing the `start` of the request, whereas in Trillian, you can get back entries from two different tiles if your `start` and `end` values cross a boundary (but only if `end-start+1 <= max_get_entries`). However, since this logic is at least as strict as Trillian's, I believe it's correct to say that it's compatible for any tile size *less than or equal to* Trillian's max_get_entries.

Preston12321 replied:
(That's not to say it's a particularly good idea to go setting these differently, because that could get really confusing really quickly)

jsha replied:
Oh interesting! Yeah, looking at the code and comments (https://github.com/google/certificate-transparency-go/blob/f6571bdcb6b6ad9972248e724ff447edf664ce45/trillian/ctfe/handlers.go#L973-L987) it appears that align_get_entries only applies to "maximally sized" get-entries. If you request smaller batches, you don't get coerced to tile boundaries.

So you're right, less than or equal is correct here. Thanks for spotting!

jsha committed:
test: remove a panic in integration test

jsha committed:
apply review feedback

jsha committed:
fix handling of CT's closed intervals

jsha committed:
doc: tile size <= max_get_entries

Trillian only applies the align_getentries tweak for "maximally sized" requests, so CTile using a tile size smaller than max_get_entries is fine.

Co-authored-by: Preston Locke <me@prestonlocke.net>

aarongable commented on .github/workflows/test.yml:
```diff
@@ -4,7 +4,7 @@ on:
     branches: [ main ]
   pull_request:
-    branches: [ main ]
+    branches: '*'
```
You want `**` so that it matches branches with `/` in their name
```suggestion
    branches: '**'
```

aarongable reviewed (approved):
Huzzah for closed intervals and off-by-one errors.

jsha committed:
test: match on ** for branches
